buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -1374,37 +1729,123 @@ public long memoryAddress() {
+         * which is currently 32). If the batch size were equal to or greater than the minimum
+         * segment count, a "batch drain" on the smallest chunks would become a "full drain,"
+         * reintroducing the original problem we're solving here.
+         * <p>
+         * The number 16 is a safe and effective choice, providing good amortization without
+         * risking this edge case or hoarding too many segments in a local cache.
+         */
+        private static final int LOCAL_CACHE_BATCH_SIZE = 16;
+        private static final int AVAILABLE = -1;
+        private static final int DEALLOCATED = Integer.MIN_VALUE;
+
+        private static final AtomicIntegerFieldUpdater<SizeClassedChunk> STATE = newUpdater(
+                SizeClassedChunk.class, "state"
+        );
+
         private static final int FREE_LIST_EMPTY = -1;
         private final int segmentSize;
-        private final MpscIntQueue freeList;
-
-        SizeClassedChunk(AbstractByteBuf delegate, Magazine magazine, boolean pooled, int segmentSize,
-                         int[] segmentOffsets, ChunkReleasePredicate shouldReleaseChunk) {
-            super(delegate, magazine, pooled, shouldReleaseChunk);
+        private final MpscIntQueue externalFreeList;
+        private final IntStack localFreeList;
+        private final Thread ownerThread;
+        private volatile int state;
+        private final int segments;
+        private boolean markedForDeallocation;
+
+        SizeClassedChunk(AbstractByteBuf delegate, AbstractMagazine magazine, int segmentSize, int[] segmentOffsets) {
+            super(delegate, magazine, true, chunkSize -> false);
+            ownerThread = magazine.ownerThread;
             this.segmentSize = segmentSize;
             int segmentCount = segmentOffsets.length;
             assert delegate.capacity() / segmentSize == segmentCount;
             assert segmentCount > 0: "Chunk must have a positive number of segments";
-            freeList = MpscIntQueue.create(segmentCount, FREE_LIST_EMPTY);
-            freeList.fill(segmentCount, new IntSupplier() {
-                int counter;
-                @Override
-                public int getAsInt() {
-                    return segmentOffsets[counter++];
-                }
-            });
+            externalFreeList = MpscIntQueue.create(segmentCount, FREE_LIST_EMPTY);
+            if (ownerThread == null) {
+                externalFreeList.fill(segmentCount, new IntSupplier() {
+                    int counter;
+
+                    @Override
+                    public int getAsInt() {
+                        return segmentOffsets[counter++];
+                    }
+                });
+                // since the allocation path for shared magazines is guarded by a lock, it's safe to use the
+                // local free list for the shared path as well.
+                localFreeList = new IntStack(segmentCount);
```

Comment:
In theory both should have the Mpsc q empty once they are created/reused and the local free list (stack) full.